content/blog/2022-05-23-vectorization/index.md

```diff
@@ -0,0 +1,238 @@
+
+i0: int = const 0;
+vecload v1 i0 a;
+i1: int = add i0 one;
+vecload v1 i1 b;
+i2: int = add i1 one;
+vecload v1 i2 a;
+i3: int = add i2 one;
+vecload v1 i3 d;
+
+i4: int = const 0;
+vecload v2 i4 b;
+i5: int = add i4 one;
+vecload v2 i5 c;
+i6: int = add i5 one;
+vecload v2 i6 c;
+i7: int = add i6 one;
+vecload v2 i7 d;
+
+sum1: vector = vecadd v1 v2;
+
+i8: int = const 0;
+w: int = vecstore sum1 i8;
+i9: int = add i8 one;
+x: int = vecstore sum1 i9;
+i10: int = add i9 one;
+y: int = vecstore sum1 i10;
+i11: int = add i10 one;
+z: int = vecstore sum1 i11;
+
+v3: vector = veczero;
+
+i12: int = const 0;
+vecload v3 i12 e;
+i13: int = add i12 one;
+vecload v3 i13 f;
+i14: int = add i13 one;
+vecload v3 i14 g;
+i15: int = add i14 one;
+vecload v3 i15 h;
+
+sum2: vector = vecadd v3 v2;
+
+i16: int = const 0;
+m: int = vecstore sum2 i16;
+i17: int = add i16 one;
+n: int = vecstore sum2 i17;
+i18: int = add i17 one;
+o: int = vecstore sum2 i18;
+i19: int = add i18 one;
+p: int = vecstore sum2 i19;
+```
+
+where only three new vectors are created, rather than four, because vector `v2` can be reused in its entirety. Note that constants like `1` are not regenerated.
+
+To deal with partial matches in opportunistic vectorization, I use the `vecmove` instruction. Consider the vector `v` associated with the quadruple (`a`, `b`, `c`, `a`). If we are required to build the vector (`a`, `b`, `d`, `a`), which differs at one location from vector `v`'s quadruple, we can reuse the vector `v` by copying it to a new vector register `v1`, and then loading `d` into the second index of `v1`. Partial matches are only considered when there are at least two matches, to lessen the number of `vecload` instructions required. Overall, using `vecmove` allows for greater reuse of vectors.
+
+To further improve opportunistic vectorization, further vector instructions could be added to allow for further vector reuse. These include shuffle or permutation instructions. This would allow for matching the quadruple of arguments (`a`, `b`, `c`, `a`) with (`a`, `a`, `b`, `c`), which is just the same vector, permuted to the right by one slot. Furthermore, permuting arguments in commutative instructions like `mul` or `add` would also allow for greater use of previously computed vectors.
+
+## Evaluation
+
+I evaluated the vectorization algorithms for correctness on my own tests, as well as the Bril Benchmarks. I restricted the benchmarks to only benchmarks using the Bril core language, as well as the Bril memory extension. The Bril speculative and Bril floating point extensions are not supported in my vectorization scheme. 
+
+Using Brench, I verified that my passes operated correctly on all the passes, compared to when my passes were not run. I also evaluated the vectorization algorithms to see how many vectors were created at run time. To do so, I instrumented the `brili-vc` interpreter to count the number of `veczero` calls that were created. The results are shown in the graph below:
```

Comment:
"on all the passes" -> "on all the benchmarks?"

Reply:
Fixed!